{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Python Datascience "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This week we are going to use jupyter lab and jupyter notebooks to run our code \n",
    "\n",
    "- previously we used \n",
    "    - `spyder` on our local machines\n",
    "    - `atom` on our local machines to remotely edit files on the server (though the `ftp-remote-edit package`) and run them through the terminal in atom that we connected to the cgat cluster using `ssh`\n",
    "    \n",
    "This week we are using `jupyter lab` and `jupyter notebooks` because we are doing some 'data science'.\n",
    "This invloves looking at your files more and plotting things and its handy to be able to look at your output in the same document that your have your code. \n",
    "\n",
    "Jupyter notebooks allow you to:\n",
    "- write text in 'markdown' format to annotate your code\n",
    "- interactively write code and see what it outputs \n",
    "- display the output right where you've run it\n",
    "- run command line statements \n",
    "\n",
    "\n",
    "\n",
    "All of the above, allow you to create a 'lab book' style document or report.\n",
    "\n",
    "\n",
    "`.ipynb` files that can be easily converted to:\n",
    "- a nice `.html` document or report thats easy to share and impress your colleagues\n",
    "- a python script - sometimes handy if your trying out code but ultimately want a `.py` file to run \n",
    "- to do this you can click on `File/Export Notebook As` then choose `executable script` for `.py` file or `html` or `pdf` etc. \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To run code in cells \n",
    "\n",
    "`Crl` + `Enter`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Cells to run code in \n",
    "Click the '+' button above"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Good practise for Notebooks \n",
    "\n",
    "- At top of notebook run: \n",
    "    - date\n",
    "    - the directory its running in \n",
    "    - your conda env \n",
    "    \n",
    "We can use bash to get all of these - using the jupyter notebook and ipython magic `!`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Feb  9 12:48:06 GMT 2021\n",
      "/ifs/obds-training/jan21/paulm/python_datascience\n",
      "obds_py3              *  /ifs/obds-training/jan21/paulm/conda/obds_conda/envs/obds_py3\n"
     ]
    }
   ],
   "source": [
    "!date\n",
    "!pwd\n",
    "! conda env list | grep '*'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Numpy \n",
    "\n",
    "- Mathematical library \n",
    "- Fast mathmatical opertations on numeric datatypes \n",
    "- Allows N-dimensional tables - all must be same type (int64, float64 etc)\n",
    "- Matrix Algebra\n",
    "- Data structure = 'Arrays' (fast access to numeric data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "x\n",
      "[[1. 2. 3.]\n",
      " [4. 5. 6.]]\n",
      "y\n",
      "[[ 7.  8.  9.]\n",
      " [10. 11. 12.]]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# import numpy library -> enable mathematical operations and create mutidimensional arrays\n",
    "import numpy as np \n",
    "\n",
    "# create 2D matrix using np.array()  \n",
    "x = np.array([[1,2,3],\n",
    "              [4,5,6]], dtype=np.float64)\n",
    "print('x')\n",
    "print(x)\n",
    "\n",
    "\n",
    "y = np.array([[7,8,9],\n",
    "              [10,11,12]], dtype=np.float64)\n",
    "print('y')\n",
    "print(y)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Can do matrix algebra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output from x+y :\n",
      "[[ 8. 10. 12.]\n",
      " [14. 16. 18.]]\n",
      "\n",
      "Can also use np.add(x,y) function:\n",
      "[[ 8. 10. 12.]\n",
      " [14. 16. 18.]]\n"
     ]
    }
   ],
   "source": [
    "# sum the elements of 2 matrices together\n",
    "print('Output from x+y :')\n",
    "print(x + y)\n",
    "print()  # this just outputs an empty line to make the output look cleaner\n",
    "\n",
    "print('Can also use np.add(x,y) function:')\n",
    "print(np.add(x,y))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Transpose (swap) rows and columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Origional x\n",
      "[[1. 2. 3.]\n",
      " [4. 5. 6.]]\n",
      "\n",
      "Transposed x\n",
      "[[1. 4.]\n",
      " [2. 5.]\n",
      " [3. 6.]]\n"
     ]
    }
   ],
   "source": [
    "# origional x\n",
    "print('Origional x')\n",
    "print(x)\n",
    "\n",
    "print()\n",
    "\n",
    "# transpose x using .T\n",
    "print(\"Transposed x\")\n",
    "print(x.T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Numpy slicing of arrays \n",
    "\n",
    "- arrayname[rows, columns] \n",
    "- remember python is 0 based (start at 0)\n",
    "- start_positon : end_possition + 1  -> get first 4 elements -> [0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3, 4]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "z = [1,2,3,4,5,6,7,8,9]\n",
    "z[0:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Origional x\n",
      "[[1. 2. 3.]\n",
      " [4. 5. 6.]]\n",
      "\n",
      "x[:2] = rows from start to 2nd possition (row 0&1)\n",
      "[[1. 2. 3.]\n",
      " [4. 5. 6.]]\n",
      "\n",
      "x[:, 1:3] = all rows, cols from start to 3rd possition (col 1-3)\n",
      "[[2. 3.]\n",
      " [5. 6.]]\n",
      "\n",
      "x[:2,1:3] = row 0 & 1 & col 1 - 2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[2., 3.],\n",
       "       [5., 6.]])"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# slicing array_name[cols, rows] ????????\n",
    "\n",
    "# origional x\n",
    "print('Origional x')\n",
    "print(x)\n",
    "\n",
    "print()\n",
    "\n",
    "print ('x[:2] = rows from start to 2nd possition (row 0&1)')\n",
    "print(x[:2])\n",
    "\n",
    "print() \n",
    "\n",
    "print ('x[:, 1:3] = all rows, cols from start to 3rd possition (col 1-3)')\n",
    "print(x[:,1:3])\n",
    "\n",
    "print() \n",
    "\n",
    "print ('x[:2,1:3] = row 0 & 1 & col 1 - 2')\n",
    "b = x[:2,1:3]\n",
    "b"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pandas = dataframes in python \n",
    "\n",
    "- built on numpy arrays\n",
    "- adds rows and column metadata & labels \n",
    "\n",
    "#### Dataframes \n",
    "\n",
    "- one datatype per column (e.g. str, int, float, list etc.)\n",
    "- Can be created from any 2D python data structure\n",
    "    - Dictionary of lists\n",
    "    - Dictionary of tuples \n",
    "    - Dictionary of series\n",
    "    - Dictionary of Dictionaries \n",
    "    - List of dictionaries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pandas Series\n",
    "\n",
    "- 1D vector built on NumPy array\n",
    "- Includes label vector - can index on this\n",
    "- Slice like a list [start_possition : stop_possition +1]\n",
    "- Can have a series of any python object (e.g. str, int, float, boolean, dict, list etc)\n",
    "- Must be all the same object type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy\n",
    "import numpy as np\n",
    "# import pandas\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create series with index or without"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      one\n",
       "1      two\n",
       "2    three\n",
       "3     four\n",
       "4     five\n",
       "5      six\n",
       "6    seven\n",
       "7    eight\n",
       "dtype: object"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pandas creates default integer index​\n",
    "s = pd.Series(['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight'])\n",
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "a   -0.703589\n",
       "b    1.506506\n",
       "c   -1.307341\n",
       "d   -0.824590\n",
       "e    0.917758\n",
       "dtype: float64"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pandas creates default integer index\n",
    "# np.random.randn(5) = make array with 5 random numbers\n",
    "s = pd.Series(np.random.randn(5), index=['a', 'b', 'c', 'd', 'e'])\n",
    "s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Print attribute of series (e.g. index or name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['a', 'b', 'c', 'd', 'e'], dtype='object')\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# index & name are attributes of Series class​\n",
    "print(s.index)\n",
    "print(s.name)  # no name currently"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Slicing series"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Origional s:\n",
      "a   -0.703589\n",
      "b    1.506506\n",
      "c   -1.307341\n",
      "d   -0.824590\n",
      "e    0.917758\n",
      "dtype: float64 \n",
      "\n",
      "s[:3] -> positions 0,1,2\n",
      "a   -0.703589\n",
      "b    1.506506\n",
      "c   -1.307341\n",
      "dtype: float64 \n",
      "\n",
      "s['a'] -> value for pos 'a'\n",
      "-0.7035889677989913 \n",
      "\n",
      "s[[4,3,1]] -> pos 4, 3, 2\n",
      "e    0.917758\n",
      "d   -0.824590\n",
      "b    1.506506\n",
      "dtype: float64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Slicing\n",
    "\n",
    "print('Origional s:')\n",
    "print(s, '\\n') # print s then an empty line\n",
    "\n",
    "print('s[:3] -> positions 0,1,2')\n",
    "\n",
    "s[:3]        # list like slicing - pos 0-2\n",
    "\n",
    "print(s[:3], '\\n')\n",
    "\n",
    "s['a']       # Index using labels\n",
    "\n",
    "print(\"s['a'] -> value for pos 'a'\")\n",
    "print(s['a'], '\\n')\n",
    "\n",
    "s[[4, 3, 1]] # array-based indexing \n",
    "\n",
    "print('s[[4,3,1]] -> pos 4, 3, 2')\n",
    "print(s[[4, 3, 1]], '\\n')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Element wide computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "origional S\n",
      "0     1\n",
      "1     2\n",
      "2     3\n",
      "3     4\n",
      "4     5\n",
      "5     6\n",
      "6     7\n",
      "7     8\n",
      "8     9\n",
      "9    10\n",
      "dtype: int64\n",
      "\n",
      "s + s\n",
      "0     2\n",
      "1     4\n",
      "2     6\n",
      "3     8\n",
      "4    10\n",
      "5    12\n",
      "6    14\n",
      "7    16\n",
      "8    18\n",
      "9    20\n",
      "dtype: int64\n",
      "\n",
      "s * 2\n",
      "0     2\n",
      "1     4\n",
      "2     6\n",
      "3     8\n",
      "4    10\n",
      "5    12\n",
      "6    14\n",
      "7    16\n",
      "8    18\n",
      "9    20\n",
      "dtype: int64\n",
      "\n"
     ]
    }
   ],
   "source": [
    "s = pd.Series([1,2,3,4,5,6,7,8,9,10])\n",
    "\n",
    "print('origional S')\n",
    "print(s)\n",
    "print()\n",
    "\n",
    "s + s # element-wise addition\n",
    "print('s + s')\n",
    "print(s+s)\n",
    "print()\n",
    "\n",
    "s * 2 # element-wise multiplication\n",
    "print('s * 2')\n",
    "print(s*2)\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create Dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From a dictionary of series (could also be a dict of lists etc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>counts</th>\n",
       "      <th>control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Actb</td>\n",
       "      <td>550</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Per1</td>\n",
       "      <td>45</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gapdh</td>\n",
       "      <td>984</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gata1</td>\n",
       "      <td>600</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Eno1</td>\n",
       "      <td>123</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   genes  counts  control\n",
       "0   Actb     550     True\n",
       "1   Per1      45    False\n",
       "2  Gapdh     984     True\n",
       "3  Gata1     600    False\n",
       "4   Eno1     123    False"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# dictionary - col names are dict 'keys', values are all in a 'list'\n",
    "dict_of_genes = {'genes': ['Actb','Per1', 'Gapdh', 'Gata1','Eno1'], \n",
    "                 'counts': [550, 45, 984, 600, 123],\n",
    "                 'control': [True, False, True, False, False]}\n",
    "\n",
    "# use df as short for dataframe\n",
    "\n",
    "df = pd.DataFrame(dict_of_genes)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'one': a    1.0\n",
      "b    2.0\n",
      "c    3.0\n",
      "dtype: float64, 'two': a    1.0\n",
      "b    2.0\n",
      "c    3.0\n",
      "d    4.0\n",
      "dtype: float64}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>one</th>\n",
       "      <th>two</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   one  two\n",
       "a  1.0  1.0\n",
       "b  2.0  2.0\n",
       "c  3.0  3.0\n",
       "d  NaN  4.0"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dataframe from dictionary of series\n",
    "d = {'one': pd.Series([1., 2., 3.], \n",
    "                      index=['a', 'b', 'c']),\n",
    "     'two': pd.Series([1., 2., 3., 4.],\n",
    "                      index=['a', 'b', 'c', 'd'])}\n",
    "\n",
    "print(d)\n",
    "\n",
    "df = pd.DataFrame(d)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Read and write Dataframes from and to files "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get Attributes of a Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['a', 'b', 'c', 'd'], dtype='object')"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.index   # row labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['one', 'two'], dtype='object')"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns # column labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4, 2)"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape   # dataframe size"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Viewing dataframes\n",
    "\n",
    "df.head()\n",
    "\n",
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get Summary of Dataframe information\n",
    " \n",
    "df.info()\n",
    "\n",
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### REad and write dataframe to file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "# Read and write to CSV\n",
    "df = pd.read_csv('file.csv')\n",
    "df.to_csv('output.csv')\n",
    "\n",
    "# tsv\n",
    "df = pd.read_csv('file.tsv', sep='\\t‘, header=0) df = pd.read_table(file_path) # tab default\n",
    "\n",
    "# Excel\n",
    "df = pd.read_excel('file.xlsx', 'Sheet1', index_col=None, na_values=['NA'])\n",
    "df.to_excel('outfile.xlsx',sheet_name='Sheet1')\n",
    "\n",
    "# HDF5\n",
    "df = pd.read_hdf('file.h5', 'df')\n",
    "df.to_hdf('foo.h5', 'df')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Add columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>one</th>\n",
       "      <th>two</th>\n",
       "      <th>three</th>\n",
       "      <th>flag</th>\n",
       "      <th>four</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   one  two  three   flag  four\n",
       "a  1.0  1.0    1.0  False   1.0\n",
       "b  2.0  2.0    4.0  False   1.0\n",
       "c  3.0  3.0    9.0   True   1.0\n",
       "d  NaN  4.0    NaN  False   NaN"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Adding columns\n",
    "df['three'] = df['one'] * df['two']\n",
    "df['flag'] = df['one'] > 2\n",
    "\n",
    "# Assign – (like dplyr mutate)\n",
    "df = df.assign(four = df['one'] /df['two'])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Delete Columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>one</th>\n",
       "      <th>two</th>\n",
       "      <th>three</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   one  two  three   flag\n",
       "a  1.0  1.0    1.0  False\n",
       "b  2.0  2.0    4.0  False\n",
       "c  3.0  3.0    9.0   True\n",
       "d  NaN  4.0    NaN  False"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.drop('four',axis=1) # need to save this to a variable to take effect, or use inplace=True in parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a    1.0\n",
      "b    1.0\n",
      "c    1.0\n",
      "d    NaN\n",
      "Name: four, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "four = df.pop('four')\n",
    "print(four)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Dataframe slicing, selecting and subsetting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "a    1.0\n",
       "b    2.0\n",
       "c    3.0\n",
       "d    NaN\n",
       "Name: one, dtype: float64"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select single column\n",
    "df['one']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>two</th>\n",
       "      <th>one</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   two  one\n",
       "a  1.0  1.0\n",
       "b  2.0  2.0\n",
       "c  3.0  3.0\n",
       "d  4.0  NaN"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select multiple columns (list)\n",
    "df[['two', 'one']] # any order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>one</th>\n",
       "      <th>two</th>\n",
       "      <th>three</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   one  two  three   flag\n",
       "a  1.0  1.0    1.0  False\n",
       "b  2.0  2.0    4.0  False\n",
       "c  3.0  3.0    9.0   True"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Slice data frame [rows]\n",
    "df[:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "one          2\n",
       "two          2\n",
       "three        4\n",
       "flag     False\n",
       "Name: b, dtype: object"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get row by label -> returns series \n",
    "df.loc['b',:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>one</th>\n",
       "      <th>two</th>\n",
       "      <th>three</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   one  two  three\n",
       "c  3.0  3.0    9.0\n",
       "d  NaN  4.0    NaN"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get row by label and then specific columns \n",
    "df.loc['c':, 'one':'three']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>three</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   three\n",
       "b    4.0\n",
       "c    9.0"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get row by interger location \n",
    "df.iloc[2:]\n",
    "df.iloc[1:3, 2:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Boolean Indexing -> returns a dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>one</th>\n",
       "      <th>two</th>\n",
       "      <th>three</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   one  two  three   flag\n",
       "b  2.0  2.0    4.0  False\n",
       "c  3.0  3.0    9.0   True"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select rows on column values\n",
    "df[df.one > 1]    # can also do df[df['one'] > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>one</th>\n",
       "      <th>two</th>\n",
       "      <th>three</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   one  two  three  flag\n",
       "a  NaN  NaN    NaN   NaN\n",
       "b  NaN  NaN    4.0   NaN\n",
       "c  3.0  3.0    9.0   NaN\n",
       "d  NaN  4.0    NaN   NaN"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select values on condition\n",
    "df[df > 2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1) Read in a idxstats table you have generated for one of your samples on the course and save to variable called `idxdf`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>195471971</td>\n",
       "      <td>7505663</td>\n",
       "      <td>323256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>130694993</td>\n",
       "      <td>4880751</td>\n",
       "      <td>248716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>122082543</td>\n",
       "      <td>7452789</td>\n",
       "      <td>385535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120129022</td>\n",
       "      <td>3104950</td>\n",
       "      <td>145924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>120421639</td>\n",
       "      <td>3222176</td>\n",
       "      <td>150791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr14</td>\n",
       "      <td>124902244</td>\n",
       "      <td>3975495</td>\n",
       "      <td>191228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr15</td>\n",
       "      <td>104043685</td>\n",
       "      <td>3796740</td>\n",
       "      <td>194000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98207768</td>\n",
       "      <td>2481586</td>\n",
       "      <td>122178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>94987271</td>\n",
       "      <td>5794590</td>\n",
       "      <td>280652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr18</td>\n",
       "      <td>90702639</td>\n",
       "      <td>2167389</td>\n",
       "      <td>103901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr19</td>\n",
       "      <td>61431566</td>\n",
       "      <td>3573941</td>\n",
       "      <td>183566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr2</td>\n",
       "      <td>182113224</td>\n",
       "      <td>6596073</td>\n",
       "      <td>320953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr3</td>\n",
       "      <td>160039680</td>\n",
       "      <td>4062539</td>\n",
       "      <td>197868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr4</td>\n",
       "      <td>156508116</td>\n",
       "      <td>5264332</td>\n",
       "      <td>262146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr5</td>\n",
       "      <td>151834684</td>\n",
       "      <td>5440642</td>\n",
       "      <td>263428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr6</td>\n",
       "      <td>149736546</td>\n",
       "      <td>4683974</td>\n",
       "      <td>223154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr7</td>\n",
       "      <td>145441459</td>\n",
       "      <td>7687960</td>\n",
       "      <td>361209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr8</td>\n",
       "      <td>129401213</td>\n",
       "      <td>4345076</td>\n",
       "      <td>212528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr9</td>\n",
       "      <td>124595110</td>\n",
       "      <td>5147071</td>\n",
       "      <td>247008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16299</td>\n",
       "      <td>2421340</td>\n",
       "      <td>80238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrX</td>\n",
       "      <td>171031299</td>\n",
       "      <td>3431992</td>\n",
       "      <td>151970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrY</td>\n",
       "      <td>91744698</td>\n",
       "      <td>310467</td>\n",
       "      <td>15700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2305942</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0          1        2        3\n",
       "0    chr1  195471971  7505663   323256\n",
       "1   chr10  130694993  4880751   248716\n",
       "2   chr11  122082543  7452789   385535\n",
       "3   chr12  120129022  3104950   145924\n",
       "4   chr13  120421639  3222176   150791\n",
       "5   chr14  124902244  3975495   191228\n",
       "6   chr15  104043685  3796740   194000\n",
       "7   chr16   98207768  2481586   122178\n",
       "8   chr17   94987271  5794590   280652\n",
       "9   chr18   90702639  2167389   103901\n",
       "10  chr19   61431566  3573941   183566\n",
       "11   chr2  182113224  6596073   320953\n",
       "12   chr3  160039680  4062539   197868\n",
       "13   chr4  156508116  5264332   262146\n",
       "14   chr5  151834684  5440642   263428\n",
       "15   chr6  149736546  4683974   223154\n",
       "16   chr7  145441459  7687960   361209\n",
       "17   chr8  129401213  4345076   212528\n",
       "18   chr9  124595110  5147071   247008\n",
       "19   chrM      16299  2421340    80238\n",
       "20   chrX  171031299  3431992   151970\n",
       "21   chrY   91744698   310467    15700\n",
       "22      *          0        0  2305942"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idxdf = pd.read_csv(\"/ifs/obds-training/jan21/exercises/rnaseq/room1/sam/ERR1755082.idxstats\", sep = '\\t', header = None)\n",
    "idxdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2) Assign column names to table -> see http://www.htslib.org/doc/samtools-idxstats.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>mapped_reads</th>\n",
       "      <th>unmapped_reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>195471971</td>\n",
       "      <td>7505663</td>\n",
       "      <td>323256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>130694993</td>\n",
       "      <td>4880751</td>\n",
       "      <td>248716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>122082543</td>\n",
       "      <td>7452789</td>\n",
       "      <td>385535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>120129022</td>\n",
       "      <td>3104950</td>\n",
       "      <td>145924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>120421639</td>\n",
       "      <td>3222176</td>\n",
       "      <td>150791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr14</td>\n",
       "      <td>124902244</td>\n",
       "      <td>3975495</td>\n",
       "      <td>191228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr15</td>\n",
       "      <td>104043685</td>\n",
       "      <td>3796740</td>\n",
       "      <td>194000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98207768</td>\n",
       "      <td>2481586</td>\n",
       "      <td>122178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>94987271</td>\n",
       "      <td>5794590</td>\n",
       "      <td>280652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr18</td>\n",
       "      <td>90702639</td>\n",
       "      <td>2167389</td>\n",
       "      <td>103901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr19</td>\n",
       "      <td>61431566</td>\n",
       "      <td>3573941</td>\n",
       "      <td>183566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr2</td>\n",
       "      <td>182113224</td>\n",
       "      <td>6596073</td>\n",
       "      <td>320953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr3</td>\n",
       "      <td>160039680</td>\n",
       "      <td>4062539</td>\n",
       "      <td>197868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr4</td>\n",
       "      <td>156508116</td>\n",
       "      <td>5264332</td>\n",
       "      <td>262146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr5</td>\n",
       "      <td>151834684</td>\n",
       "      <td>5440642</td>\n",
       "      <td>263428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr6</td>\n",
       "      <td>149736546</td>\n",
       "      <td>4683974</td>\n",
       "      <td>223154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr7</td>\n",
       "      <td>145441459</td>\n",
       "      <td>7687960</td>\n",
       "      <td>361209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr8</td>\n",
       "      <td>129401213</td>\n",
       "      <td>4345076</td>\n",
       "      <td>212528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr9</td>\n",
       "      <td>124595110</td>\n",
       "      <td>5147071</td>\n",
       "      <td>247008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrM</td>\n",
       "      <td>16299</td>\n",
       "      <td>2421340</td>\n",
       "      <td>80238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrX</td>\n",
       "      <td>171031299</td>\n",
       "      <td>3431992</td>\n",
       "      <td>151970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chrY</td>\n",
       "      <td>91744698</td>\n",
       "      <td>310467</td>\n",
       "      <td>15700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2305942</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seq_name  seq_length  mapped_reads  unmapped_reads\n",
       "0      chr1   195471971       7505663          323256\n",
       "1     chr10   130694993       4880751          248716\n",
       "2     chr11   122082543       7452789          385535\n",
       "3     chr12   120129022       3104950          145924\n",
       "4     chr13   120421639       3222176          150791\n",
       "5     chr14   124902244       3975495          191228\n",
       "6     chr15   104043685       3796740          194000\n",
       "7     chr16    98207768       2481586          122178\n",
       "8     chr17    94987271       5794590          280652\n",
       "9     chr18    90702639       2167389          103901\n",
       "10    chr19    61431566       3573941          183566\n",
       "11     chr2   182113224       6596073          320953\n",
       "12     chr3   160039680       4062539          197868\n",
       "13     chr4   156508116       5264332          262146\n",
       "14     chr5   151834684       5440642          263428\n",
       "15     chr6   149736546       4683974          223154\n",
       "16     chr7   145441459       7687960          361209\n",
       "17     chr8   129401213       4345076          212528\n",
       "18     chr9   124595110       5147071          247008\n",
       "19     chrM       16299       2421340           80238\n",
       "20     chrX   171031299       3431992          151970\n",
       "21     chrY    91744698        310467           15700\n",
       "22        *           0             0         2305942"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idxdf.columns = [\"seq_name\", \"seq_length\", \"mapped_reads\", \"unmapped_reads\"]\n",
    "idxdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3)Apply the `describe()` and `info()` methods to the `idxdf` object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         seq_length  mapped_reads  unmapped_reads\n",
      "count  2.300000e+01  2.300000e+01    2.300000e+01\n",
      "mean   1.185016e+08  4.232502e+06    3.031257e+05\n",
      "std    4.911871e+07  2.051681e+06    4.457848e+05\n",
      "min    0.000000e+00  0.000000e+00    1.570000e+04\n",
      "25%    9.659752e+07  3.163563e+06    1.513805e+05\n",
      "50%    1.245951e+08  4.062539e+06    2.125280e+05\n",
      "75%    1.507856e+08  5.352487e+06    2.720400e+05\n",
      "max    1.954720e+08  7.687960e+06    2.305942e+06\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 23 entries, 0 to 22\n",
      "Data columns (total 4 columns):\n",
      " #   Column          Non-Null Count  Dtype \n",
      "---  ------          --------------  ----- \n",
      " 0   seq_name        23 non-null     object\n",
      " 1   seq_length      23 non-null     int64 \n",
      " 2   mapped_reads    23 non-null     int64 \n",
      " 3   unmapped_reads  23 non-null     int64 \n",
      "dtypes: int64(3), object(1)\n",
      "memory usage: 864.0+ bytes\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 8 entries, count to max\n",
      "Data columns (total 3 columns):\n",
      " #   Column          Non-Null Count  Dtype  \n",
      "---  ------          --------------  -----  \n",
      " 0   seq_length      8 non-null      float64\n",
      " 1   mapped_reads    8 non-null      float64\n",
      " 2   unmapped_reads  8 non-null      float64\n",
      "dtypes: float64(3)\n",
      "memory usage: 256.0+ bytes\n",
      "None\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 23 entries, 0 to 22\n",
      "Data columns (total 4 columns):\n",
      " #   Column          Non-Null Count  Dtype \n",
      "---  ------          --------------  ----- \n",
      " 0   seq_name        23 non-null     object\n",
      " 1   seq_length      23 non-null     int64 \n",
      " 2   mapped_reads    23 non-null     int64 \n",
      " 3   unmapped_reads  23 non-null     int64 \n",
      "dtypes: int64(3), object(1)\n",
      "memory usage: 864.0+ bytes\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'describe'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-94-87b0398e149d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0midxdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0midxdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdescribe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0midxdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minfo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdescribe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'describe'"
     ]
    }
   ],
   "source": [
    "print(idxdf.describe())\n",
    "print(idxdf.info())\n",
    "print(idxdf.describe().info())\n",
    "print(idxdf.info().describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4)Look at the `shape` and `index` attributes of `idxdf`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shape_idxdf = idxdf.shape\n",
    "print(shape_idxdf)\n",
    "print(idxdf.index)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5) From `idxdf` select rows 6-12 and create a new variable called `idxdf2` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "idxdf2 = idxdf[6:13]\n",
    "idxdf2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6) From `idxdf` make a new dataframe `chr_mapped` consisting of columns 1&3 (chr & mapped). Save to a new file called `mapped_reads_per_chr.tsv` - make sure the output is tab seperated​\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_mapped = idxdf[['seq_name','mapped_reads']]\n",
    "display(seq_mapped.to_csv('mapped_reads_per_chr.tsv', sep=\"\\t\", header= True, index = False))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "7) Create a new column for idxdf called `reads_per_base` by dividing number of `mapped reads` by `length` column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "idxdf['reads_per_base'] = idxdf['mapped_reads']/idxdf['seq_length']\n",
    "print(idxdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "8) Subset the rows that have `reads_per_base` values greater than 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "readsperbase_over5 = idxdf[idxdf['reads_per_base'] > 0.05]\n",
    "readsperbase_over5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "9)Create a new column of length / 1 million to give a `length_Mbs` column (length in Megabases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "idxdf['length_Mbs'] = idxdf['seq_length']/1000000\n",
    "display(idxdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "10) Subset the rows that have `reads_per_base` values greater than 0.05 and a chromosome length under 100 Mb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset1 = idxdf[ (idxdf['reads_per_base'] > 0.05) & (idxdf['length_Mbs'] < 100)]\n",
    "subset1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "11) Delete the `reads_per_base` column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "idxdf.drop('reads_per_base', axis = 1, inplace = True)\n",
    "display(idxdf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sorting Dataframes "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "# Sort by single column\n",
    "df.sort_values(by=['col1'])\n",
    "\n",
    "# Sort by multiple columns\n",
    "df.sort_values(by=['col1', 'col2'])\n",
    "\n",
    "# Descending order\n",
    "df.sort_values(by='col1', ascending=False)\n",
    "\n",
    "# NAs first\n",
    "df.sort_values(by='col1', ascending=False, na_position='first')\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Operations \n",
    "```\n",
    "# Element-wise mathematical operations\n",
    "df * 5\n",
    "df **3\n",
    "\n",
    "# Boolean logic\n",
    "df1 & df2\n",
    "\n",
    "# Numpy functions\n",
    "np.log(df)\n",
    "\n",
    "# Statistical summary\n",
    "df.describe()\n",
    "\n",
    "# statistical functions\n",
    "df.mean() # per column\n",
    "df.mean(axis=1) # per row\n",
    "\n",
    "# Matrix algebra\n",
    "df.T # transpose of matrix \n",
    "df.dot(df) # matrix multiplication\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Apply Functions along an axis\n",
    "\n",
    "```\n",
    "# apply to entire dataframe\n",
    "df.apply(np.sqrt)\n",
    "\n",
    "# apply to cols\n",
    "df.apply(np.sum, axis=0)\n",
    "\n",
    "\n",
    "# apply to rows\n",
    "df.apply(np.sum, axis=1)\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Missing Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "# drop rows with nan\n",
    "df.dropna(how='any')\n",
    "\n",
    "# Replace nan with specific value\n",
    "df.fillna(value=5)\n",
    "\n",
    "# Create Boolean mask\n",
    "pd.isna(df)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Categorical Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "df[\"biotype\"] = df[\"raw_biotype\"].astype(\"category\")\n",
    "df[\"biotype\"].cat.categories = \"protein_coding\", \"rRNA\", \"lincRNA\"]\n",
    "\n",
    "\n",
    "# Show distinct categories are present in dataset\n",
    "df[\"biotype\"].unique()\n",
    "\n",
    "# Number of distinct categories \n",
    "df[\"biotype\"].nunique()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Grouping\n",
    "\n",
    "```\n",
    "df.groupby('A').sum()\n",
    "\n",
    "df.groupby(['A','C']).count().reset\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### String manipulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_names = {'samples' : ['mm_Chd7WTatac_R1','mm_Chd7WTatac_R2','mm_Chd7WTatac_R3','mm_Chd7KOatac_R1','mm_Chd7KOatac_R2','mm_Chd7KOatac_R3'],\n",
    "                'file_type' : ['fastq.gz','fastq.gz','fastq.gz','fastq.gz','fastq.gz','fastq.gz']}\n",
    "df = pd.DataFrame(sample_names)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['samples'].str.split('_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['samples'].str.split('_',expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['replicate'] = df['samples'].str.split('_').str[-1]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Add column and set a value depending on another column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['samples'].str.contains('WT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['condition'] = np.where(df['samples'].str.contains('WT'), 'WT', 'KO')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combining Datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "See documentation here: https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html\n",
    "\n",
    "```\n",
    "# Concatenate rows\n",
    "pd.concat([df[:3], df[3:7], df[7:]])​\n",
    "\n",
    "# Merge (inner by default)\n",
    "result = pd.merge(df1, df2, on='key')\n",
    "result = pd.merge(df1, df2, on='key', how='left')\n",
    "\n",
    "# Join on index\n",
    "result = df1.join(df2, how='outer')\n",
    "\n",
    "# Join on key\n",
    "result = df1.join(df2, on='key')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1) read in ensembl gtf files to create `gtf`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/ifs/obds-training/jan21/paulm/conda/obds_conda/envs/obds_py3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3072: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "gtf = pd.read_csv(\"/ifs/obds-training/jan21/exercises/pandas/Mus_musculus.GRCm38.102.chr.gtf.gz\", sep = '\\t', header = None, comment = \"#\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2) Assign appropriate column names https://www.ensembl.org/info/website/upload/gff.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3102016</td>\n",
       "      <td>3102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname   source     feature    start      end score strand frame  \\\n",
       "0       1   havana        gene  3073253  3074322     .      +     .   \n",
       "1       1   havana  transcript  3073253  3074322     .      +     .   \n",
       "2       1   havana        exon  3073253  3074322     .      +     .   \n",
       "3       1  ensembl        gene  3102016  3102125     .      +     .   \n",
       "4       1  ensembl  transcript  3102016  3102125     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSMUSG00000102693\"; gene_version \"1\"...  \n",
       "1  gene_id \"ENSMUSG00000102693\"; gene_version \"1\"...  \n",
       "2  gene_id \"ENSMUSG00000102693\"; gene_version \"1\"...  \n",
       "3  gene_id \"ENSMUSG00000064842\"; gene_version \"1\"...  \n",
       "4  gene_id \"ENSMUSG00000064842\"; gene_version \"1\"...  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "gtf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3) Find the unique values for 'source' column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['havana', 'ensembl', 'ensembl_havana', 'mirbase', 'insdc'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf['source'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4) Subset all values from 'ensembl_havana' and assign to new df called 'ens_df'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3214482</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671348</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>3671346</td>\n",
       "      <td>3671348</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868399</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>61175993</td>\n",
       "      <td>61175995</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868400</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>61175459</td>\n",
       "      <td>61175461</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868401</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>61175996</td>\n",
       "      <td>61176309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868402</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>61174686</td>\n",
       "      <td>61175458</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868407</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61223957</td>\n",
       "      <td>61228429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>558494 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source          feature     start       end score  \\\n",
       "6             1  ensembl_havana             gene   3205901   3671498     .   \n",
       "13            1  ensembl_havana       transcript   3214482   3671498     .   \n",
       "14            1  ensembl_havana             exon   3670552   3671498     .   \n",
       "15            1  ensembl_havana              CDS   3670552   3671348     .   \n",
       "16            1  ensembl_havana      start_codon   3671346   3671348     .   \n",
       "...         ...             ...              ...       ...       ...   ...   \n",
       "1868399      19  ensembl_havana      start_codon  61175993  61175995     .   \n",
       "1868400      19  ensembl_havana       stop_codon  61175459  61175461     .   \n",
       "1868401      19  ensembl_havana   five_prime_utr  61175996  61176309     .   \n",
       "1868402      19  ensembl_havana  three_prime_utr  61174686  61175458     .   \n",
       "1868407      19  ensembl_havana             gene  61223957  61228429     .   \n",
       "\n",
       "        strand frame                                          attribute  \n",
       "6            -     .  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...  \n",
       "13           -     .  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...  \n",
       "14           -     .  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...  \n",
       "15           -     0  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...  \n",
       "16           -     0  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...  \n",
       "...        ...   ...                                                ...  \n",
       "1868399      -     0  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...  \n",
       "1868400      -     0  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...  \n",
       "1868401      -     .  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...  \n",
       "1868402      -     .  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...  \n",
       "1868407      -     .  gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...  \n",
       "\n",
       "[558494 rows x 9 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df = gtf[gtf['source'] == 'ensembl_havana']\n",
    "ens_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5) What is the size of ens_df?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(558494, 9)"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df.shape\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6) What is the size of ens_df?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(558494, 9)"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "7) Group ens_df by 'feature' column & apply the 'count' function to get the number of rows in each feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>15498367263615</td>\n",
       "      <td>15498401175865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Selenocysteine</th>\n",
       "      <td>1711915680</td>\n",
       "      <td>1711915726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>16527565840806</td>\n",
       "      <td>16527631853102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_utr</th>\n",
       "      <td>2280181473232</td>\n",
       "      <td>2280185942977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>1681144030164</td>\n",
       "      <td>1682202771711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>1611251981576</td>\n",
       "      <td>1611252024044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>1596835675694</td>\n",
       "      <td>1596835717827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>three_prime_utr</th>\n",
       "      <td>1648153114159</td>\n",
       "      <td>1648179394081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>1687760689816</td>\n",
       "      <td>1688743752355</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          start             end\n",
       "feature                                        \n",
       "CDS              15498367263615  15498401175865\n",
       "Selenocysteine       1711915680      1711915726\n",
       "exon             16527565840806  16527631853102\n",
       "five_prime_utr    2280181473232   2280185942977\n",
       "gene              1681144030164   1682202771711\n",
       "start_codon       1611251981576   1611252024044\n",
       "stop_codon        1596835675694   1596835717827\n",
       "three_prime_utr   1648153114159   1648179394081\n",
       "transcript        1687760689816   1688743752355"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>203373</td>\n",
       "      <td>203373</td>\n",
       "      <td>203373</td>\n",
       "      <td>203373</td>\n",
       "      <td>203373</td>\n",
       "      <td>203373</td>\n",
       "      <td>203373</td>\n",
       "      <td>203373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Selenocysteine</th>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>216470</td>\n",
       "      <td>216470</td>\n",
       "      <td>216470</td>\n",
       "      <td>216470</td>\n",
       "      <td>216470</td>\n",
       "      <td>216470</td>\n",
       "      <td>216470</td>\n",
       "      <td>216470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_utr</th>\n",
       "      <td>29833</td>\n",
       "      <td>29833</td>\n",
       "      <td>29833</td>\n",
       "      <td>29833</td>\n",
       "      <td>29833</td>\n",
       "      <td>29833</td>\n",
       "      <td>29833</td>\n",
       "      <td>29833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>22436</td>\n",
       "      <td>22436</td>\n",
       "      <td>22436</td>\n",
       "      <td>22436</td>\n",
       "      <td>22436</td>\n",
       "      <td>22436</td>\n",
       "      <td>22436</td>\n",
       "      <td>22436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>21291</td>\n",
       "      <td>21291</td>\n",
       "      <td>21291</td>\n",
       "      <td>21291</td>\n",
       "      <td>21291</td>\n",
       "      <td>21291</td>\n",
       "      <td>21291</td>\n",
       "      <td>21291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>21101</td>\n",
       "      <td>21101</td>\n",
       "      <td>21101</td>\n",
       "      <td>21101</td>\n",
       "      <td>21101</td>\n",
       "      <td>21101</td>\n",
       "      <td>21101</td>\n",
       "      <td>21101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>three_prime_utr</th>\n",
       "      <td>21645</td>\n",
       "      <td>21645</td>\n",
       "      <td>21645</td>\n",
       "      <td>21645</td>\n",
       "      <td>21645</td>\n",
       "      <td>21645</td>\n",
       "      <td>21645</td>\n",
       "      <td>21645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>22322</td>\n",
       "      <td>22322</td>\n",
       "      <td>22322</td>\n",
       "      <td>22322</td>\n",
       "      <td>22322</td>\n",
       "      <td>22322</td>\n",
       "      <td>22322</td>\n",
       "      <td>22322</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seqname  source   start     end   score  strand   frame  \\\n",
       "feature                                                                    \n",
       "CDS               203373  203373  203373  203373  203373  203373  203373   \n",
       "Selenocysteine        23      23      23      23      23      23      23   \n",
       "exon              216470  216470  216470  216470  216470  216470  216470   \n",
       "five_prime_utr     29833   29833   29833   29833   29833   29833   29833   \n",
       "gene               22436   22436   22436   22436   22436   22436   22436   \n",
       "start_codon        21291   21291   21291   21291   21291   21291   21291   \n",
       "stop_codon         21101   21101   21101   21101   21101   21101   21101   \n",
       "three_prime_utr    21645   21645   21645   21645   21645   21645   21645   \n",
       "transcript         22322   22322   22322   22322   22322   22322   22322   \n",
       "\n",
       "                 attribute  \n",
       "feature                     \n",
       "CDS                 203373  \n",
       "Selenocysteine          23  \n",
       "exon                216470  \n",
       "five_prime_utr       29833  \n",
       "gene                 22436  \n",
       "start_codon          21291  \n",
       "stop_codon           21101  \n",
       "three_prime_utr      21645  \n",
       "transcript           22322  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(ens_df.groupby('feature').sum())\n",
    "display(ens_df.groupby('feature').count())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "8) Create a new df called 'genes' by subsetting all rows with 'gene' as a feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3999557</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025900\"; gene_version \"13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4490931</td>\n",
       "      <td>4497354</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025902\"; gene_version \"13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4773206</td>\n",
       "      <td>4785739</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000033845\"; gene_version \"13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807788</td>\n",
       "      <td>4848410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903\"; gene_version \"14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868189</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60864051</td>\n",
       "      <td>60874556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868209</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60889749</td>\n",
       "      <td>61095968</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868272</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61053840</td>\n",
       "      <td>61140840</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000074733\"; gene_version \"14...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868395</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61174686</td>\n",
       "      <td>61176309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868407</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61223957</td>\n",
       "      <td>61228429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22436 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "6             1  ensembl_havana    gene   3205901   3671498     .      -   \n",
       "70            1  ensembl_havana    gene   3999557   4409241     .      -   \n",
       "160           1  ensembl_havana    gene   4490931   4497354     .      -   \n",
       "283           1  ensembl_havana    gene   4773206   4785739     .      -   \n",
       "354           1  ensembl_havana    gene   4807788   4848410     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "1868189      19  ensembl_havana    gene  60864051  60874556     .      -   \n",
       "1868209      19  ensembl_havana    gene  60889749  61095968     .      +   \n",
       "1868272      19  ensembl_havana    gene  61053840  61140840     .      -   \n",
       "1868395      19  ensembl_havana    gene  61174686  61176309     .      -   \n",
       "1868407      19  ensembl_havana    gene  61223957  61228429     .      -   \n",
       "\n",
       "        frame                                          attribute  \n",
       "6           .  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...  \n",
       "70          .  gene_id \"ENSMUSG00000025900\"; gene_version \"13...  \n",
       "160         .  gene_id \"ENSMUSG00000025902\"; gene_version \"13...  \n",
       "283         .  gene_id \"ENSMUSG00000033845\"; gene_version \"13...  \n",
       "354         .  gene_id \"ENSMUSG00000025903\"; gene_version \"14...  \n",
       "...       ...                                                ...  \n",
       "1868189     .  gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...  \n",
       "1868209     .  gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...  \n",
       "1868272     .  gene_id \"ENSMUSG00000074733\"; gene_version \"14...  \n",
       "1868395     .  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...  \n",
       "1868407     .  gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...  \n",
       "\n",
       "[22436 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_gtf = ens_df[ens_df['feature'] == 'gene'].copy()\n",
    "display(genes_gtf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "9) Create a new column called 'gene_length' and calculate the length of each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "      <td>465597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3999557</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025900\"; gene_version \"13...</td>\n",
       "      <td>409684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4490931</td>\n",
       "      <td>4497354</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025902\"; gene_version \"13...</td>\n",
       "      <td>6423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4773206</td>\n",
       "      <td>4785739</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000033845\"; gene_version \"13...</td>\n",
       "      <td>12533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807788</td>\n",
       "      <td>4848410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903\"; gene_version \"14...</td>\n",
       "      <td>40622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868189</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60864051</td>\n",
       "      <td>60874556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...</td>\n",
       "      <td>10505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868209</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60889749</td>\n",
       "      <td>61095968</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...</td>\n",
       "      <td>206219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868272</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61053840</td>\n",
       "      <td>61140840</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000074733\"; gene_version \"14...</td>\n",
       "      <td>87000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868395</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61174686</td>\n",
       "      <td>61176309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "      <td>1623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868407</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61223957</td>\n",
       "      <td>61228429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...</td>\n",
       "      <td>4472</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22436 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "6             1  ensembl_havana    gene   3205901   3671498     .      -   \n",
       "70            1  ensembl_havana    gene   3999557   4409241     .      -   \n",
       "160           1  ensembl_havana    gene   4490931   4497354     .      -   \n",
       "283           1  ensembl_havana    gene   4773206   4785739     .      -   \n",
       "354           1  ensembl_havana    gene   4807788   4848410     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "1868189      19  ensembl_havana    gene  60864051  60874556     .      -   \n",
       "1868209      19  ensembl_havana    gene  60889749  61095968     .      +   \n",
       "1868272      19  ensembl_havana    gene  61053840  61140840     .      -   \n",
       "1868395      19  ensembl_havana    gene  61174686  61176309     .      -   \n",
       "1868407      19  ensembl_havana    gene  61223957  61228429     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "6           .  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...       465597  \n",
       "70          .  gene_id \"ENSMUSG00000025900\"; gene_version \"13...       409684  \n",
       "160         .  gene_id \"ENSMUSG00000025902\"; gene_version \"13...         6423  \n",
       "283         .  gene_id \"ENSMUSG00000033845\"; gene_version \"13...        12533  \n",
       "354         .  gene_id \"ENSMUSG00000025903\"; gene_version \"14...        40622  \n",
       "...       ...                                                ...          ...  \n",
       "1868189     .  gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...        10505  \n",
       "1868209     .  gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...       206219  \n",
       "1868272     .  gene_id \"ENSMUSG00000074733\"; gene_version \"14...        87000  \n",
       "1868395     .  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...         1623  \n",
       "1868407     .  gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...         4472  \n",
       "\n",
       "[22436 rows x 10 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_gtf[\"gene_length\"] = genes_gtf[\"end\"] - genes_gtf[\"start\"]\n",
    "genes_gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "10) Calculate the mean and median of gene lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>2.243600e+04</td>\n",
       "      <td>2.243600e+04</td>\n",
       "      <td>2.243600e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.493065e+07</td>\n",
       "      <td>7.497784e+07</td>\n",
       "      <td>4.718941e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>4.429087e+07</td>\n",
       "      <td>4.428967e+07</td>\n",
       "      <td>1.095276e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>7.251280e+05</td>\n",
       "      <td>7.974090e+05</td>\n",
       "      <td>9.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.671332e+07</td>\n",
       "      <td>3.674402e+07</td>\n",
       "      <td>5.905750e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>7.336766e+07</td>\n",
       "      <td>7.340731e+07</td>\n",
       "      <td>1.630850e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.080928e+08</td>\n",
       "      <td>1.081357e+08</td>\n",
       "      <td>4.379100e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.951368e+08</td>\n",
       "      <td>1.951767e+08</td>\n",
       "      <td>2.960899e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              start           end   gene_length\n",
       "count  2.243600e+04  2.243600e+04  2.243600e+04\n",
       "mean   7.493065e+07  7.497784e+07  4.718941e+04\n",
       "std    4.429087e+07  4.428967e+07  1.095276e+05\n",
       "min    7.251280e+05  7.974090e+05  9.000000e+00\n",
       "25%    3.671332e+07  3.674402e+07  5.905750e+03\n",
       "50%    7.336766e+07  7.340731e+07  1.630850e+04\n",
       "75%    1.080928e+08  1.081357e+08  4.379100e+04\n",
       "max    1.951368e+08  1.951767e+08  2.960899e+06"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "16308.5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "47189.4075147085"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(genes_gtf.describe())\n",
    "display(np.median(genes_gtf['gene_length']))\n",
    "display(genes_gtf['gene_length'].mean())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "11) Find the top 10 longest genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_genes_by_length = genes_gtf.sort_values(by='gene_length', ascending = False).head(10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "12) Find the lowest 20 shortest genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1516168</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113464761</td>\n",
       "      <td>113464770</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000093876\"; gene_version \"1\"...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516217</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113526800</td>\n",
       "      <td>113526809</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000095549\"; gene_version \"1\"...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516144</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113454942</td>\n",
       "      <td>113454951</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094957\"; gene_version \"1\"...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516180</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113469426</td>\n",
       "      <td>113469435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000096396\"; gene_version \"1\"...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516156</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113460101</td>\n",
       "      <td>113460110</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000096884\"; gene_version \"1\"...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1119803</th>\n",
       "      <td>14</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>54113468</td>\n",
       "      <td>54113478</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000096749\"; gene_version \"2\"...</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516126</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113430528</td>\n",
       "      <td>113430538</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094028\"; gene_version \"1\"...</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>671211</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>41533201</td>\n",
       "      <td>41533212</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000095668\"; gene_version \"1\"...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516134</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113449588</td>\n",
       "      <td>113449599</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000096420\"; gene_version \"2\"...</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>671266</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>41542163</td>\n",
       "      <td>41542176</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094569\"; gene_version \"1\"...</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516130</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113448214</td>\n",
       "      <td>113448229</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094552\"; gene_version \"1\"...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1119812</th>\n",
       "      <td>14</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>54122226</td>\n",
       "      <td>54122241</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000096176\"; gene_version \"1\"...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516213</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113525313</td>\n",
       "      <td>113525329</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000093818\"; gene_version \"1\"...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516194</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113475400</td>\n",
       "      <td>113475416</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000096568\"; gene_version \"1\"...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516140</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113450851</td>\n",
       "      <td>113450867</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000095656\"; gene_version \"1\"...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516184</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113470694</td>\n",
       "      <td>113470710</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000095444\"; gene_version \"1\"...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516160</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113461369</td>\n",
       "      <td>113461385</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000096250\"; gene_version \"1\"...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516148</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113456720</td>\n",
       "      <td>113456736</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094057\"; gene_version \"1\"...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516172</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113466027</td>\n",
       "      <td>113466043</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000095897\"; gene_version \"1\"...</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1516220</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>113528032</td>\n",
       "      <td>113528054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000076632\"; gene_version \"1\"...</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "1516168      12  ensembl_havana    gene  113464761  113464770     .      -   \n",
       "1516217      12  ensembl_havana    gene  113526800  113526809     .      -   \n",
       "1516144      12  ensembl_havana    gene  113454942  113454951     .      -   \n",
       "1516180      12  ensembl_havana    gene  113469426  113469435     .      -   \n",
       "1516156      12  ensembl_havana    gene  113460101  113460110     .      -   \n",
       "1119803      14  ensembl_havana    gene   54113468   54113478     .      +   \n",
       "1516126      12  ensembl_havana    gene  113430528  113430538     .      -   \n",
       "671211        6  ensembl_havana    gene   41533201   41533212     .      +   \n",
       "1516134      12  ensembl_havana    gene  113449588  113449599     .      -   \n",
       "671266        6  ensembl_havana    gene   41542163   41542176     .      +   \n",
       "1516130      12  ensembl_havana    gene  113448214  113448229     .      -   \n",
       "1119812      14  ensembl_havana    gene   54122226   54122241     .      +   \n",
       "1516213      12  ensembl_havana    gene  113525313  113525329     .      -   \n",
       "1516194      12  ensembl_havana    gene  113475400  113475416     .      -   \n",
       "1516140      12  ensembl_havana    gene  113450851  113450867     .      -   \n",
       "1516184      12  ensembl_havana    gene  113470694  113470710     .      -   \n",
       "1516160      12  ensembl_havana    gene  113461369  113461385     .      -   \n",
       "1516148      12  ensembl_havana    gene  113456720  113456736     .      -   \n",
       "1516172      12  ensembl_havana    gene  113466027  113466043     .      -   \n",
       "1516220      12  ensembl_havana    gene  113528032  113528054     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "1516168     .  gene_id \"ENSMUSG00000093876\"; gene_version \"1\"...            9  \n",
       "1516217     .  gene_id \"ENSMUSG00000095549\"; gene_version \"1\"...            9  \n",
       "1516144     .  gene_id \"ENSMUSG00000094957\"; gene_version \"1\"...            9  \n",
       "1516180     .  gene_id \"ENSMUSG00000096396\"; gene_version \"1\"...            9  \n",
       "1516156     .  gene_id \"ENSMUSG00000096884\"; gene_version \"1\"...            9  \n",
       "1119803     .  gene_id \"ENSMUSG00000096749\"; gene_version \"2\"...           10  \n",
       "1516126     .  gene_id \"ENSMUSG00000094028\"; gene_version \"1\"...           10  \n",
       "671211      .  gene_id \"ENSMUSG00000095668\"; gene_version \"1\"...           11  \n",
       "1516134     .  gene_id \"ENSMUSG00000096420\"; gene_version \"2\"...           11  \n",
       "671266      .  gene_id \"ENSMUSG00000094569\"; gene_version \"1\"...           13  \n",
       "1516130     .  gene_id \"ENSMUSG00000094552\"; gene_version \"1\"...           15  \n",
       "1119812     .  gene_id \"ENSMUSG00000096176\"; gene_version \"1\"...           15  \n",
       "1516213     .  gene_id \"ENSMUSG00000093818\"; gene_version \"1\"...           16  \n",
       "1516194     .  gene_id \"ENSMUSG00000096568\"; gene_version \"1\"...           16  \n",
       "1516140     .  gene_id \"ENSMUSG00000095656\"; gene_version \"1\"...           16  \n",
       "1516184     .  gene_id \"ENSMUSG00000095444\"; gene_version \"1\"...           16  \n",
       "1516160     .  gene_id \"ENSMUSG00000096250\"; gene_version \"1\"...           16  \n",
       "1516148     .  gene_id \"ENSMUSG00000094057\"; gene_version \"1\"...           16  \n",
       "1516172     .  gene_id \"ENSMUSG00000095897\"; gene_version \"1\"...           16  \n",
       "1516220     .  gene_id \"ENSMUSG00000076632\"; gene_version \"1\"...           22  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sorted_genes_by_length = genes_gtf.sort_values(by='gene_length', ascending = True).head(20)\n",
    "display(sorted_genes_by_length)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "13) Split attributes column on ';' and create new dataframe, merge this onto the ens_gtf dataframe using pandas merge ​"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...</td>\n",
       "      <td>465597</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951\"</td>\n",
       "      <td>gene_version \"5\"</td>\n",
       "      <td>gene_name \"Xkr4\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000026353\"</td>\n",
       "      <td>havana_gene_version \"2\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3999557</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025900\"; gene_version \"13...</td>\n",
       "      <td>409684</td>\n",
       "      <td>gene_id \"ENSMUSG00000025900\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"Rp1\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000049985\"</td>\n",
       "      <td>havana_gene_version \"3\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4490931</td>\n",
       "      <td>4497354</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025902\"; gene_version \"13...</td>\n",
       "      <td>6423</td>\n",
       "      <td>gene_id \"ENSMUSG00000025902\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"Sox17\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000050014\"</td>\n",
       "      <td>havana_gene_version \"7\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4773206</td>\n",
       "      <td>4785739</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000033845\"; gene_version \"13...</td>\n",
       "      <td>12533</td>\n",
       "      <td>gene_id \"ENSMUSG00000033845\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"Mrpl15\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000029329\"</td>\n",
       "      <td>havana_gene_version \"3\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>4807788</td>\n",
       "      <td>4848410</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903\"; gene_version \"14...</td>\n",
       "      <td>40622</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903\"</td>\n",
       "      <td>gene_version \"14\"</td>\n",
       "      <td>gene_name \"Lypla1\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000021562\"</td>\n",
       "      <td>havana_gene_version \"4\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868189</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60864051</td>\n",
       "      <td>60874556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...</td>\n",
       "      <td>10505</td>\n",
       "      <td>gene_id \"ENSMUSG00000024997\"</td>\n",
       "      <td>gene_version \"7\"</td>\n",
       "      <td>gene_name \"Prdx3\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000021212\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868209</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>60889749</td>\n",
       "      <td>61095968</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...</td>\n",
       "      <td>206219</td>\n",
       "      <td>gene_id \"ENSMUSG00000003228\"</td>\n",
       "      <td>gene_version \"9\"</td>\n",
       "      <td>gene_name \"Grk5\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000074399\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868272</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61053840</td>\n",
       "      <td>61140840</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000074733\"; gene_version \"14...</td>\n",
       "      <td>87000</td>\n",
       "      <td>gene_id \"ENSMUSG00000074733\"</td>\n",
       "      <td>gene_version \"14\"</td>\n",
       "      <td>gene_name \"Zfp950\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000033031\"</td>\n",
       "      <td>havana_gene_version \"5\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868395</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61174686</td>\n",
       "      <td>61176309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...</td>\n",
       "      <td>1623</td>\n",
       "      <td>gene_id \"ENSMUSG00000094649\"</td>\n",
       "      <td>gene_version \"1\"</td>\n",
       "      <td>gene_name \"Gm7102\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000042892\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1868407</th>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>61223957</td>\n",
       "      <td>61228429</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...</td>\n",
       "      <td>4472</td>\n",
       "      <td>gene_id \"ENSMUSG00000059326\"</td>\n",
       "      <td>gene_version \"7\"</td>\n",
       "      <td>gene_name \"Csf2ra\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td>havana_gene \"OTTMUSG00000074402\"</td>\n",
       "      <td>havana_gene_version \"1\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22436 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "6             1  ensembl_havana    gene   3205901   3671498     .      -   \n",
       "70            1  ensembl_havana    gene   3999557   4409241     .      -   \n",
       "160           1  ensembl_havana    gene   4490931   4497354     .      -   \n",
       "283           1  ensembl_havana    gene   4773206   4785739     .      -   \n",
       "354           1  ensembl_havana    gene   4807788   4848410     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "1868189      19  ensembl_havana    gene  60864051  60874556     .      -   \n",
       "1868209      19  ensembl_havana    gene  60889749  61095968     .      +   \n",
       "1868272      19  ensembl_havana    gene  61053840  61140840     .      -   \n",
       "1868395      19  ensembl_havana    gene  61174686  61176309     .      -   \n",
       "1868407      19  ensembl_havana    gene  61223957  61228429     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \\\n",
       "6           .  gene_id \"ENSMUSG00000051951\"; gene_version \"5\"...       465597   \n",
       "70          .  gene_id \"ENSMUSG00000025900\"; gene_version \"13...       409684   \n",
       "160         .  gene_id \"ENSMUSG00000025902\"; gene_version \"13...         6423   \n",
       "283         .  gene_id \"ENSMUSG00000033845\"; gene_version \"13...        12533   \n",
       "354         .  gene_id \"ENSMUSG00000025903\"; gene_version \"14...        40622   \n",
       "...       ...                                                ...          ...   \n",
       "1868189     .  gene_id \"ENSMUSG00000024997\"; gene_version \"7\"...        10505   \n",
       "1868209     .  gene_id \"ENSMUSG00000003228\"; gene_version \"9\"...       206219   \n",
       "1868272     .  gene_id \"ENSMUSG00000074733\"; gene_version \"14...        87000   \n",
       "1868395     .  gene_id \"ENSMUSG00000094649\"; gene_version \"1\"...         1623   \n",
       "1868407     .  gene_id \"ENSMUSG00000059326\"; gene_version \"7\"...         4472   \n",
       "\n",
       "                                    0                   1  \\\n",
       "6        gene_id \"ENSMUSG00000051951\"    gene_version \"5\"   \n",
       "70       gene_id \"ENSMUSG00000025900\"   gene_version \"13\"   \n",
       "160      gene_id \"ENSMUSG00000025902\"   gene_version \"13\"   \n",
       "283      gene_id \"ENSMUSG00000033845\"   gene_version \"13\"   \n",
       "354      gene_id \"ENSMUSG00000025903\"   gene_version \"14\"   \n",
       "...                               ...                 ...   \n",
       "1868189  gene_id \"ENSMUSG00000024997\"    gene_version \"7\"   \n",
       "1868209  gene_id \"ENSMUSG00000003228\"    gene_version \"9\"   \n",
       "1868272  gene_id \"ENSMUSG00000074733\"   gene_version \"14\"   \n",
       "1868395  gene_id \"ENSMUSG00000094649\"    gene_version \"1\"   \n",
       "1868407  gene_id \"ENSMUSG00000059326\"    gene_version \"7\"   \n",
       "\n",
       "                           2                              3  \\\n",
       "6           gene_name \"Xkr4\"   gene_source \"ensembl_havana\"   \n",
       "70           gene_name \"Rp1\"   gene_source \"ensembl_havana\"   \n",
       "160        gene_name \"Sox17\"   gene_source \"ensembl_havana\"   \n",
       "283       gene_name \"Mrpl15\"   gene_source \"ensembl_havana\"   \n",
       "354       gene_name \"Lypla1\"   gene_source \"ensembl_havana\"   \n",
       "...                      ...                            ...   \n",
       "1868189    gene_name \"Prdx3\"   gene_source \"ensembl_havana\"   \n",
       "1868209     gene_name \"Grk5\"   gene_source \"ensembl_havana\"   \n",
       "1868272   gene_name \"Zfp950\"   gene_source \"ensembl_havana\"   \n",
       "1868395   gene_name \"Gm7102\"   gene_source \"ensembl_havana\"   \n",
       "1868407   gene_name \"Csf2ra\"   gene_source \"ensembl_havana\"   \n",
       "\n",
       "                                      4                                  5  \\\n",
       "6         gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000026353\"   \n",
       "70        gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000049985\"   \n",
       "160       gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000050014\"   \n",
       "283       gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000029329\"   \n",
       "354       gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000021562\"   \n",
       "...                                 ...                                ...   \n",
       "1868189   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000021212\"   \n",
       "1868209   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000074399\"   \n",
       "1868272   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000033031\"   \n",
       "1868395   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000042892\"   \n",
       "1868407   gene_biotype \"protein_coding\"   havana_gene \"OTTMUSG00000074402\"   \n",
       "\n",
       "                                6 7   \n",
       "6         havana_gene_version \"2\"     \n",
       "70        havana_gene_version \"3\"     \n",
       "160       havana_gene_version \"7\"     \n",
       "283       havana_gene_version \"3\"     \n",
       "354       havana_gene_version \"4\"     \n",
       "...                           ... ..  \n",
       "1868189   havana_gene_version \"1\"     \n",
       "1868209   havana_gene_version \"1\"     \n",
       "1868272   havana_gene_version \"5\"     \n",
       "1868395   havana_gene_version \"1\"     \n",
       "1868407   havana_gene_version \"1\"     \n",
       "\n",
       "[22436 rows x 18 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "attributes = genes_gtf['attribute'].str.split(';', expand = True)\n",
    "attributes\n",
    "genes_full_attributes = pd.concat([genes_gtf, attributes], axis = 1)\n",
    "display(genes_full_attributes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extra: \n",
    "    Count the number of genes per chromosome, and add the number of genes per chomosome onto the idxdf from earlier ​"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "display(ens_df.groupby('feature').count())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
